File: framework/test2/floating/matrix/Matrix.java
import soda.unittest.*;

import java.util.*;
import java.util.function.Supplier;
import java.util.stream.*;

import soda.leetcode.*;
import soda.unittest.Validators;
import soda.unittest.work.TestWork;

import static soda.unittest.LoggerHelper.logger;

class Solution {
    public double[][] matrixMultiply(double[][] a, double[][] b) {
        int rows = a.length, cols = b[0].length;
        var res = new double[rows][cols];
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                double c = 0.0;
                for (int k = 0; k < b.length; ++k) {
                    c += a[i][k] * b[k][j];
                }
                res[i][j] = c;
            }
        }
        return res;
    }
}

public class Matrix implements Supplier<TestWork> {

    @Override
    public TestWork get() {
        var work = new TestWork(new Solution(), "matrixMultiply");
        // var work = TestWork.forStruct(Struct.class);
        // work.setValidator((e, r) -> {...});
        // work.setCompareSerial(true);
        return work;
    }

    public static void main(String[] args) throws Exception {
        new Matrix().get().run();
    }

}

File: framework/test2/floating/matrix/Matrix.scala
import soda.scala.unittest._

import scala.collection.mutable
import scala.reflect.runtime.universe.typeOf

class Solution {}
object Solution {
  def matrixMultiply(a: Array[Array[Double]], b: Array[Array[Double]]): Array[Array[Double]] = {
    val rows = a.size
    val cols = b(0).size
    val res = Array.ofDim[Double](rows, cols)
    for (i <- 0 until rows) {
      for (j <- 0 until cols) {
        var c: Double = 0.0
        for (k <- b.indices) {
          c += a(i)(k) * b(k)(j)
        }
        res(i)(j) = c
      }
    }
    res
  }
}

class Matrix {
  def get(): TestWork = {
    val work = TestWork.forObject(typeOf[Solution], "matrixMultiply")
    // val work = TestWork.forStruct(...)
    // work.setValidator((R, R) => Boolean)
    // work.compareSerial = true
    work
  }
}

object Matrix extends App {
  new Matrix().get().run()
}
